File: 2338. Count the Number of Ideal Arrays/idealArrays.js
//First solution where I use 
//=>Back Tracking Algorithm SOLUTION #1
//=> Solve 21 out of 47
var idealArrays = function (n, maxValue) {
    let ans = []
    let cnt = 0;
    const MOD = 1000000007

    function solve(arr) {
        if (arr.length === n) {
            cnt++
            return;
        }
        for (let i = 1; i <= maxValue; i++) {
            //if the length > 0
            if (arr.length > 0) {
                if (i % arr[arr.length - 1] !== 0) continue
            }
            //do
            arr.push(i)
            //recursion
            solve(arr)
            //undo
            arr.pop()
        }
    }

    solve(ans)

    return cnt % MOD
};


//Second solution where I use 
//=>Dynamic Programming SOLUTION #2
//=> Solve 35 out of 47
var idealArrays = function (n, maxValue) {
    let mp = new Map()
    const MOD = 1000000007

    function solve(idx, last) {
        let key = `${idx}-${last}`
        if (mp.has(key)) return mp.get(key);
        //base case
        if (idx === n) return 1;
        let total = 0
        for (let next = last; next <= maxValue; next += last) {
            total = (total + solve(idx + 1, next)) % MOD
        }
        mp.set(key, total)
        return total
    }
    return solve(0, 1)
};

File: 2338. Count the Number of Ideal Arrays/idealArrays.cpp
#include <map>
#include <iostream>
using namespace std;

//=>Dynamic Programming SOLUTION
//=> Solve 35 out of 47

class Solution {
public:
    map<string,int> mp;
    int MOD = 1000000007;
    int solve(int max,int n,int idx, int last){
        string key = to_string(idx) + '-' + to_string(last);
        if(mp[key])return mp[key];

        if(idx == n)return 1;
        int total = 0;
        for(int next=last; next<=max; next+=last){
            total = (total + solve(max,n,idx+1, next)) % MOD;
        }
        mp[key] = total;
        return total;
    }
    int idealArrays(int n, int maxValue) {
        
        return solve(maxValue, n,0,1);
    }
};